components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt

```diff
@@ -0,0 +1,139 @@
+     * @param downloadListener a listener to be notify when the download is completed.
+     * @param refererURL the url from where this download was referred.
+     * @param cookie any additional cookie to add as part of the download request.
+     * @return the id reference of the scheduled download.
+     */
+    @RequiresPermission(allOf = arrayOf(INTERNET, WRITE_EXTERNAL_STORAGE))
+    fun download(
+        download: Download,
+        downloadListener: OnDownloadFinish = {},
+        refererURL: String = "",
+        cookie: String = ""
+    ): Long {
+
+        if (!context.isPermissionGranted(INTERNET, WRITE_EXTERNAL_STORAGE))
+            throw SecurityException("You must be granted INTERNET and WRITE_EXTERNAL_STORAGE permissions")
+
+        androidDownloadManager = context.getSystemService(DOWNLOAD_SERVICE) as AndroidDownloadManager
+
+        val fileName = getFileName(download)
+
+        val request = Request(Uri.parse(download.url))
+            .setNotificationVisibility(VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
+            .setMimeType(download.contentType)
+            .addRequestHeader("User-Agent", download.userAgent)
+
+        if (refererURL.isNotEmpty())
+            request.addRequestHeader("Cookie", cookie)
+
+        val downloadID: Long
+
+        request.setDestinationInExternalPublicDir(download.destinationDirectory, fileName)
+
+        downloadID = androidDownloadManager.enqueue(request)
+        queuedDownloads[downloadID] = downloadListener
+
+        if (!isSubscribedReceiver)
+            registerBroadcastReceiver(context)
+
+        return downloadID
+    }
+
+    /**
+     * Use for register a listener for scheduled download.
+     * in case you did not provide one on [download] function.
+     * @param downloadID the id that reference to the schedule download.
+     * @param listener a listener to be notify when the download is completed.
+     */
+    @Suppress("unused")
+    fun registerListener(downloadID: Long, listener: OnDownloadFinish) {
+        if (downloadID in queuedDownloads)
+            queuedDownloads[downloadID] = listener
+        else
+            throw IllegalArgumentException("Unable to find $downloadID on the queued downloads")
+    }
+
+    /**
+     * Remove all the listeners.
+     */
+    fun unregisterListeners() {
+        context.unregisterReceiver(onDownloadComplete)
+        isSubscribedReceiver = false
+        queuedDownloads.clear()
+    }
+
+    private fun registerBroadcastReceiver(context: Context) {
+        val filter = IntentFilter(ACTION_DOWNLOAD_COMPLETE)
+        context.registerReceiver(onDownloadComplete, filter)
+        isSubscribedReceiver = true
+    }
+
+    private fun getFileName(download: Download): String? {
+        return if (!download.fileName.isNullOrEmpty())
+            download.fileName
+        else
+            DownloadUtils.guessFileName(
+                "",
+                download.url,
+                download.contentType
+            )
```

Comment:
I think we should file a follow-up to make sure fileName is never null. No reason to let consumers call `guessFileName`. And I think in `browser-engine-system` we already do that. We just need to make sure that the gecko-flavors do that too.

Reply:
I'm a bit confused, can you help me to clarify please.

1.Do you want that I change Download.fileName property to Non-null? **e.g fileName: String** instead of  **fileName: String?** ?

2.Or are you referring to `private fun getFileName(download: Download): String?` type return  **String?**

Reply:
Option 1: We should always come up with a file name. But that's something for a separate issue.

Reply:
https://github.com/mozilla-mobile/android-components/issues/1017